_posts/2023-12-20-standard-libraries.markdown

```diff
@@ -0,0 +1,407 @@
+  // ...
+}
+```
+
+How do we find the first and last digits?
+
+## Sequences and Wrappers
+
+Having a look at the [top-level index of Dafny standard libraries](https://github.com/dafny-lang/dafny/tree/master/Source/DafnyStandardLibraries),
+we notice there's a `Std.Collections` library with [a submodule for `Seq`](https://github.com/dafny-lang/dafny/blob/master/Source/DafnyStandardLibraries/src/Std/Collections/Seq.dfy), and sure enough, there are functions like `IndexOf` and `LastIndexOf`. Perfect!
+
+Or perhaps you were distracted by the `Std.Strings` module, where you might expect these utilities to live.
+But in Dafny, strings are really nothing more than sequences of characters:
+the type `string` is just an alias for `seq<char>`.
+This means many typical string operations are provided as general sequence operations instead.
+Don't worry, we'll get to that library in a moment.
+
+Let's consider for a moment how you might want to use a function like `IndexOf` in general.
+Finding the index of an element in a sequence will certainly not always succeed,
+since the sequence might not contain the element.
+In most standard libraries, a function like `IndexOf` may return an invalid sentinel index value,
+like the length of the sequence or `-`.
+But often you actually know the element is in the sequence
+because of the way the rest of the program is structured.
+Perhaps you just added the element to a list and then sorted it.
+Because of the invariants of sorting, you know the element is still in there.
+The great thing about Dafny is that it can actually understand this kind of reasoning,
+and know when it is safe to assume an operation will succeed.
+
+Therefore a common pattern in the Dafny standard libraries
+is to have two different versions of a function or method:
+a partial version with preconditions,
+and a complete version that may not succeed.
+The `Std.Wrappers` library provides a few extremely common datatypes for this:
+`Option<T>`, for a value that may not exist,
+and `Result<T, E>`, for the result of an operation that may fail with an error value instead.
+The `Seq` library and most of the other libraries use these datatypes all over the place
+in order to define partial operations.
+
+In our specific case, there is both an `IndexOf` function,
+which requires the element is present in the sequence and returns a `nat` that is always a valid index in the sequence,
+and `IndexOfOption`, which produces an `Option<nat>` instead of a `nat`.
+The latter is the better match for our situation,
+since we're eventually going to read our puzzle input from a file on disk,
+and we don't actually **know** that every line contains at least one digit.
+
+Note that in general we could get by with only the partial version of all these operations,
+since we could always just explicitly check the precondition first
+before invoking the operation.
+The downside of that approach, though, is that there's a performance cost:
```

Comment:
Yeah, I feel like what I said is still true (currently) and simpler. It also seems like the two-pass code would not be that straightforward to optimize in this case (not impossible but not as easy as something like `sequence.Filter(...).Map(...)`), so I wouldn't want to imply that it will definitely be compiled to something as fast some day:

```dafny
if IsNumberStr(s) then
  OfNat(s)
else
  None
```

Reply:
Yes, I realize that I had misread this; now I see that you're advocating for matching on an optional rather than testing a predicate and then running a total operation, which I agree with.